File: tests/DesktopGrass.Win2D.Tests/SimTests/BunnyTests.cs
// BunnyTests.cs - §18 Bunny critter tests.
// Mirrors tests/DesktopGrass.Native.Tests/src/bunny_tests.cpp.

using System;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class BunnyTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, Monitor1920, Constants.DEFAULT_DENSITY),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, Monitor1920);
        sim.ResetEntities(seed);
        return sim;
    }

    private static Sim BuildGrassSim(ulong seed = Constants.CANONICAL_TEST_SEED)
    {
        var sim = BuildSim(seed);
        sim.SetScene(Scene.Grass);
        return sim;
    }

    private static int CountKind(Sim sim, EntityKind kind) => sim.Entities.Count(e => e.Kind == kind);

    private static Entity BunnyEntity(double x = 500.0, double vx = Constants.BUNNY_HOP_SPEED_MIN)
    {
        return new Entity
        {
            Kind = EntityKind.Bunny,
            Size = Constants.BUNNY_BODY_RADIUS,
            X = x,
            Y = Constants.STRIP_HEIGHT + Constants.HEADROOM - Constants.BUNNY_BODY_HEIGHT - Constants.BUNNY_LEG_LENGTH,
            Vx = vx,
            RotationSpeed = Math.Abs(vx),
            Lifetime = -1.0,
            State = Constants.BUNNY_STATE_HOPPING,
            StateTimer = Constants.BUNNY_HOP_DURATION,
        };
    }

    private static int PrngCount(ref Prng side, int minCount, int maxCount)
    {
        double draw = side.Uniform(minCount, maxCount + 1);
        int count = (int)Math.Floor(draw);
        if (count < minCount) count = minCount;
        if (count > maxCount) count = maxCount;
        return count;
    }

    private static void AdvanceSheep(ref Prng side, int count)
    {
        for (int i = 0; i < count; i++)
        {
            double margin = Constants.SHEEP_BODY_RADIUS + 8.0;
            _ = side.Uniform(margin, Monitor1920 - margin);
            _ = side.Uniform(Constants.SHEEP_WALK_SPEED_MIN, Constants.SHEEP_WALK_SPEED_MAX);
            _ = side.Uniform(0.0, 1.0);
            _ = side.NextU32();
            _ = side.Uniform(Constants.SHEEP_WALK_DURATION_MIN, Constants.SHEEP_WALK_DURATION_MAX);
            _ = side.Index((uint)Constants.SHEEP_NAME_POOL.Length);
        }
    }

    private static void AdvanceCats(ref Prng side, int count)
    {
        for (int i = 0; i < count; i++)
        {
            double margin = Constants.CAT_BODY_RADIUS + 8.0;
            _ = side.Uniform(margin, Monitor1920 - margin);
            _ = side.Uniform(Constants.CAT_WALK_SPEED_MIN, Constants.CAT_WALK_SPEED_MAX);
            _ = side.Uniform(0.0, 1.0);
            _ = side.NextU32();
            _ = side.Uniform(Constants.CAT_WALK_DURATION_MIN, Constants.CAT_WALK_DURATION_MAX);
            _ = side.Index((uint)Constants.CAT_NAME_POOL.Length);
            _ = side.Index((uint)Constants.CAT_COAT_VARIANT_COUNT);
        }
    }

    [Fact]
    public void BunnyConstantsArePinnedToSpecValues()
    {
        Assert.Equal(1, Constants.BUNNY_COUNT_MIN);
        Assert.Equal(2, Constants.BUNNY_COUNT_MAX);
        Assert.Equal(22.0, Constants.BUNNY_HOP_SPEED_MIN);
        Assert.Equal(38.0, Constants.BUNNY_HOP_SPEED_MAX);
        Assert.Equal(8.0, Constants.BUNNY_BODY_RADIUS);
        Assert.Equal(6.5, Constants.BUNNY_BODY_HEIGHT);
        Assert.Equal(4.2, Constants.BUNNY_HEAD_RADIUS);
        Assert.Equal(9.0, Constants.BUNNY_EAR_HEIGHT);
        Assert.Equal(2.2, Constants.BUNNY_EAR_WIDTH);
        Assert.Equal(3.0, Constants.BUNNY_EAR_SPACING);
        Assert.Equal(4.0, Constants.BUNNY_LEG_LENGTH);
        Assert.Equal(2.4, Constants.BUNNY_TAIL_RADIUS);
        Assert.Equal(0xFF8A6A4Au, Constants.BUNNY_BODY_COLOR);
        Assert.Equal(0xFFC4A98Du, Constants.BUNNY_BELLY_COLOR);
        Assert.Equal(0xFF8A6A4Au, Constants.BUNNY_EAR_COLOR);
        Assert.Equal(0xFFD9A0A0u, Constants.BUNNY_EAR_INNER_COLOR);
        Assert.Equal(0xFFF7F4EBu, Constants.BUNNY_TAIL_COLOR);
        Assert.Equal(0xFF1A1208u, Constants.BUNNY_EYE_COLOR);
        Assert.Equal(0xFF8A4040u, Constants.BUNNY_NOSE_COLOR);
        Assert.Equal((byte)0, Constants.BUNNY_STATE_HOPPING);
        Assert.Equal((byte)1, Constants.BUNNY_STATE_GRAZING);
        Assert.Equal((byte)2, Constants.BUNNY_STATE_IDLE);
        Assert.Equal((byte)3, Constants.BUNNY_STATE_SLEEPING);
        Assert.Equal((byte)4, Constants.BUNNY_STATE_STARTLED);
        Assert.Equal(0.40, Constants.BUNNY_HOP_DURATION);
        Assert.Equal(8.0, Constants.BUNNY_HOP_HEIGHT);
        Assert.Equal(0.05, Constants.BUNNY_HOP_GAP_MIN);
        Assert.Equal(0.20, Constants.BUNNY_HOP_GAP_MAX);
        Assert.Equal(2.5, Constants.BUNNY_GRAZE_DURATION_MIN);
        Assert.Equal(4.5, Constants.BUNNY_GRAZE_DURATION_MAX);
        Assert.Equal(2.0, Constants.BUNNY_IDLE_DURATION_MIN);
        Assert.Equal(4.0, Constants.BUNNY_IDLE_DURATION_MAX);
        Assert.Equal(6.0, Constants.BUNNY_SLEEP_DURATION_MIN);
        Assert.Equal(12.0, Constants.BUNNY_SLEEP_DURATION_MAX);
        Assert.Equal(0.55, Constants.BUNNY_GRAZE_PROBABILITY);
        Assert.Equal(0.30, Constants.BUNNY_IDLE_PROBABILITY);
        Assert.Equal(0.05, Constants.BUNNY_SLEEP_PROB_DAY);
        Assert.Equal(0.40, Constants.BUNNY_SLEEP_PROB_NIGHT);
        Assert.Equal(90.0, Constants.BUNNY_STARTLE_RADIUS);
        Assert.Equal(2.0, Constants.BUNNY_STARTLE_BOOST);
        Assert.Equal(14.0, Constants.BUNNY_STARTLE_HOP_HEIGHT);
        Assert.Equal(3.0, Constants.BUNNY_STARTLE_DURATION);
        Assert.Equal(6.0, Constants.BUNNY_NOSE_TWITCH_FREQ);
        Assert.Equal(0.5, Constants.BUNNY_NOSE_TWITCH_AMP);
        Assert.Equal(1.2, Constants.BUNNY_EAR_WIGGLE_FREQ);
        Assert.Equal(0.20, Constants.BUNNY_EAR_WIGGLE_AMP);
        Assert.Equal(Constants.SHEEP_ZZZ_CYCLE_SEC, Constants.BUNNY_ZZZ_CYCLE_SEC);
        Assert.Equal(Constants.SHEEP_ZZZ_RISE * 0.7, Constants.BUNNY_ZZZ_RISE);
        Assert.Equal(Constants.SHEEP_ZZZ_SIZE_START * 0.7, Constants.BUNNY_ZZZ_SIZE_START);
        Assert.Equal(Constants.SHEEP_ZZZ_SIZE_END * 0.7, Constants.BUNNY_ZZZ_SIZE_END);
        Assert.Equal(12, Constants.BUNNY_NAME_POOL.Length);
        Assert.Equal("Clover", Constants.BUNNY_NAME_POOL[0]);
        Assert.Equal("Snowdrop", Constants.BUNNY_NAME_POOL[11]);
    }

    [Fact]
    public void GrassGenerationProducesBunnyCountInRange()
    {
        for (ulong i = 0; i < 128; i++)
        {
            ulong seed = unchecked(Constants.CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15UL);
            var sim = BuildGrassSim(seed);
            Assert.InRange(CountKind(sim, EntityKind.Bunny), Constants.BUNNY_COUNT_MIN, Constants.BUNNY_COUNT_MAX);
        }
    }

    [Fact]
    public void BunniesAreGrassSceneOnly()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);
        Assert.Equal(0, CountKind(sim, EntityKind.Bunny));
        sim.SetScene(Scene.Winter);
        Assert.Equal(0, CountKind(sim, EntityKind.Bunny));
        sim.SetCritter(CritterKind.Bunny);
        Assert.Equal(0, CountKind(sim, EntityKind.Bunny));
    }

    [Fact]
    public void GeneratedBunniesHaveSpeedRange()
    {
        var sim = BuildGrassSim();
        foreach (var e in sim.Entities.Where(e => e.Kind == EntityKind.Bunny))
        {
            Assert.InRange(Math.Abs(e.Vx), Constants.BUNNY_HOP_SPEED_MIN, Constants.BUNNY_HOP_SPEED_MAX);
            Assert.Equal(Math.Abs(e.Vx), e.RotationSpeed, 9);
        }
    }

    [Fact]
    public void GeneratedBunniesHaveNamesInPool()
    {
        var sim = BuildGrassSim();
        foreach (var e in sim.Entities.Where(e => e.Kind == EntityKind.Bunny))
        {
            Assert.Contains(Constants.BUNNY_NAME_POOL[e.NameIndex], Constants.BUNNY_NAME_POOL);
        }
    }

    [Fact]
    public void BunnyPrngDrawOrderFollowsSheepAndCats()
    {
        var side = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.CRITTER_PRNG_SALT);
        var sim = BuildGrassSim();

        int sheepCount = PrngCount(ref side, Constants.SHEEP_COUNT_MIN, Constants.SHEEP_COUNT_MAX);
        AdvanceSheep(ref side, sheepCount);
        int catCount = PrngCount(ref side, Constants.CAT_COUNT_MIN, Constants.CAT_COUNT_MAX);
        AdvanceCats(ref side, catCount);
        int bunnyCount = PrngCount(ref side, Constants.BUNNY_COUNT_MIN, Constants.BUNNY_COUNT_MAX);
        Assert.Equal(bunnyCount, CountKind(sim, EntityKind.Bunny));

        int seen = 0;
        foreach (var e in sim.Entities.Where(e => e.Kind == EntityKind.Bunny))
        {
            double margin = Constants.BUNNY_BODY_RADIUS + 8.0;
            double xFrac = side.Uniform(0.0, 1.0);
            double expectedX = margin + xFrac * (Monitor1920 - 2.0 * margin);
            ulong vxSign = side.NextU64() & 1UL;
            double expectedDir = vxSign != 0UL ? 1.0 : -1.0;
            double expectedSpeed = side.Uniform(Constants.BUNNY_HOP_SPEED_MIN, Constants.BUNNY_HOP_SPEED_MAX);
            byte expectedName = (byte)side.Index((uint)Constants.BUNNY_NAME_POOL.Length);

            Assert.Equal(expectedX, e.X, 9);
            Assert.Equal(expectedDir * expectedSpeed, e.Vx, 9);
            Assert.Equal(expectedName, e.NameIndex);
            seen++;
        }
        Assert.Equal(bunnyCount, seen);
    }

    [Fact]
    public void BunnyEdgeBounceFlipsDirection()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        sim.Entities.Clear();
        var e = BunnyEntity(Monitor1920 - (Constants.BUNNY_BODY_RADIUS + 2.0) + 0.1,
                            Constants.BUNNY_HOP_SPEED_MIN);
        e.StateTimer = 10.0;
        sim.Entities.Add(e);

        sim.TickEntities(0.016);

        Assert.True(sim.Entities[0].Vx < 0.0);
    }

    [Fact]
    public void BunnyStartleRadiusHopsAwayAndOutsideClickDoesNothing()
    {
        var sim = BuildSim();
        sim.Entities.Clear();
        var e = BunnyEntity(500.0, -Constants.BUNNY_HOP_SPEED_MIN);
        e.State = Constants.BUNNY_STATE_IDLE;
        e.StateTimer = 3.0;
        sim.Entities.Add(e);

        sim.ApplyClick(e.X - 20.0, e.Y, 0.0);
        Assert.Equal(Constants.BUNNY_STATE_STARTLED, sim.Entities[0].State);
        Assert.True(sim.Entities[0].Vx > 0.0);
        Assert.Equal(Constants.BUNNY_STARTLE_DURATION, sim.Entities[0].StateTimer, 9);

        var after = sim.Entities[0];
        double vxBefore = after.Vx;
        byte stateBefore = after.State;
        sim.ApplyClick(after.X + Constants.BUNNY_STARTLE_RADIUS + 10.0, after.Y, 0.0);
        Assert.Equal(stateBefore, sim.Entities[0].State);
        Assert.Equal(vxBefore, sim.Entities[0].Vx, 9);
    }

    [Fact]
    public void BunnyWakesFromSleepOnStartle()
    {
        var sim = BuildSim();
        sim.Entities.Clear();
        var e = BunnyEntity(500.0, Constants.BUNNY_HOP_SPEED_MIN);
        e.State = Constants.BUNNY_STATE_SLEEPING;
        e.StateTimer = 10.0;
        sim.Entities.Add(e);

        sim.ApplyClick(e.X + 10.0, e.Y, 0.0);

        Assert.Equal(Constants.BUNNY_STATE_STARTLED, sim.Entities[0].State);
        Assert.NotEqual(Constants.BUNNY_STATE_SLEEPING, sim.Entities[0].State);
        Assert.True(sim.Entities[0].Vx < 0.0);
    }

    [Fact]
    public void BunnyHopArcIsBounded()
    {
        Assert.Equal(0.0, Sim.BunnyHopYOffset(0.0, startled: false), 12);
        Assert.Equal(0.0, Sim.BunnyHopYOffset(Constants.BUNNY_HOP_DURATION, startled: false), 12);
        double peak = Sim.BunnyHopYOffset(Constants.BUNNY_HOP_DURATION * 0.5, startled: false);
        Assert.True(peak > 0.0);
        Assert.True(peak <= Constants.BUNNY_HOP_HEIGHT);
    }

    [Fact]
    public void BunnyStateTransitionProbabilitiesAreStable()
    {
        var p = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.CRITTER_PRNG_SALT);
        const int n = 10000;
        int graze = 0;
        int idle = 0;
        int sleep = 0;
        for (int i = 0; i < n; i++)
        {
            byte state = Sim.BunnyChooseRestState(ref p, 12);
            if (state == Constants.BUNNY_STATE_GRAZING) graze++;
            else if (state == Constants.BUNNY_STATE_IDLE) idle++;
            else if (state == Constants.BUNNY_STATE_SLEEPING) sleep++;
        }

        double sleepProb = Constants.BUNNY_SLEEP_PROB_DAY;
        double activeWeight = Constants.BUNNY_GRAZE_PROBABILITY + Constants.BUNNY_IDLE_PROBABILITY;
        double expectedGraze = (1.0 - sleepProb) * Constants.BUNNY_GRAZE_PROBABILITY / activeWeight;
        double expectedIdle = (1.0 - sleepProb) * Constants.BUNNY_IDLE_PROBABILITY / activeWeight;
        Assert.InRange(sleep / (double)n, sleepProb - 0.02, sleepProb + 0.02);
        Assert.InRange(graze / (double)n, expectedGraze - 0.02, expectedGraze + 0.02);
        Assert.InRange(idle / (double)n, expectedIdle - 0.02, expectedIdle + 0.02);
    }

    [Fact]
    public void BunnyTimeOfDaySleepBiasIsDayNight()
    {
        Assert.Equal(Constants.BUNNY_SLEEP_PROB_DAY, Sim.BunnySleepProbForLocalHour(12));
        Assert.Equal(Constants.BUNNY_SLEEP_PROB_NIGHT, Sim.BunnySleepProbForLocalHour(0));

        const int n = 10000;
        var noon = Prng.Init(Constants.CANONICAL_TEST_SEED ^ 0x1234UL);
        var midnight = Prng.Init(Constants.CANONICAL_TEST_SEED ^ 0x5678UL);
        int noonSleep = 0;
        int midnightSleep = 0;
        for (int i = 0; i < n; i++)
        {
            if (Sim.BunnyChooseRestState(ref noon, 12) == Constants.BUNNY_STATE_SLEEPING) noonSleep++;
            if (Sim.BunnyChooseRestState(ref midnight, 0) == Constants.BUNNY_STATE_SLEEPING) midnightSleep++;
        }
        Assert.InRange(noonSleep / (double)n,
                       Constants.BUNNY_SLEEP_PROB_DAY - 0.02,
                       Constants.BUNNY_SLEEP_PROB_DAY + 0.02);
        Assert.InRange(midnightSleep / (double)n,
                       Constants.BUNNY_SLEEP_PROB_NIGHT - 0.02,
                       Constants.BUNNY_SLEEP_PROB_NIGHT + 0.02);
    }
}

File: tests/DesktopGrass.Native.Tests/src/bunny_tests.cpp
// bunny_tests.cpp
//
// §18 Bunny critter tests. Mirrors Win2D BunnyTests.cs.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>
#include <cmath>
#include <cwchar>

using namespace desktopgrass;

namespace {

constexpr double Monitor1920 = 1920.0;

int count_kind(const Sim& sim, EntityKind kind) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [kind](const Entity& e) { return e.kind == kind; }));
}

Sim build_grass_sim(uint64_t seed = CANONICAL_TEST_SEED) {
    Sim sim = sim_init(seed, Monitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Grass);
    sim_set_critter(sim, CritterKind::Bunny);
    return sim;
}

Entity bunny_entity(double x = 500.0, double vx = BUNNY_HOP_SPEED_MIN) {
    Entity e{};
    e.kind = EntityKind::Bunny;
    e.size = BUNNY_BODY_RADIUS;
    e.x = x;
    e.y = STRIP_HEIGHT + HEADROOM - BUNNY_BODY_HEIGHT - BUNNY_LEG_LENGTH;
    e.vx = vx;
    e.rotationSpeed = std::abs(vx);
    e.lifetime = -1.0;
    e.state = BUNNY_STATE_HOPPING;
    e.stateTimer = BUNNY_HOP_DURATION;
    return e;
}

InputEvent click_event(double x, double y) {
    InputEvent ev{};
    ev.type = EventType::Click;
    ev.x = x;
    ev.y = y;
    ev.time = 0.0;
    return ev;
}

int prng_count(Prng& side, int minCount, int maxCount) {
    const double draw = prng_uniform(side, static_cast<double>(minCount), static_cast<double>(maxCount + 1));
    int count = static_cast<int>(std::floor(draw));
    if (count < minCount) count = minCount;
    if (count > maxCount) count = maxCount;
    return count;
}

void advance_sheep(Prng& side, int count) {
    for (int i = 0; i < count; ++i) {
        const double margin = SHEEP_BODY_RADIUS + 8.0;
        (void)prng_uniform(side, margin, Monitor1920 - margin);
        (void)prng_uniform(side, SHEEP_WALK_SPEED_MIN, SHEEP_WALK_SPEED_MAX);
        (void)prng_uniform(side, 0.0, 1.0);
        (void)prng_next_u32(side);
        (void)prng_uniform(side, SHEEP_WALK_DURATION_MIN, SHEEP_WALK_DURATION_MAX);
        (void)prng_index(side, static_cast<uint32_t>(sizeof(SHEEP_NAME_POOL) / sizeof(SHEEP_NAME_POOL[0])));
    }
}

void advance_cats(Prng& side, int count) {
    for (int i = 0; i < count; ++i) {
        const double margin = CAT_BODY_RADIUS + 8.0;
        (void)prng_uniform(side, margin, Monitor1920 - margin);
        (void)prng_uniform(side, CAT_WALK_SPEED_MIN, CAT_WALK_SPEED_MAX);
        (void)prng_uniform(side, 0.0, 1.0);
        (void)prng_next_u32(side);
        (void)prng_uniform(side, CAT_WALK_DURATION_MIN, CAT_WALK_DURATION_MAX);
        (void)prng_index(side, static_cast<uint32_t>(sizeof(CAT_NAME_POOL) / sizeof(CAT_NAME_POOL[0])));
        (void)prng_index(side, static_cast<uint32_t>(CAT_COAT_VARIANT_COUNT));
    }
}

bool bunny_name_in_pool(const Entity& e) {
    if (e.nameIndex >= sizeof(BUNNY_NAME_POOL) / sizeof(BUNNY_NAME_POOL[0])) return false;
    const wchar_t* name = BUNNY_NAME_POOL[e.nameIndex];
    for (const wchar_t* candidate : BUNNY_NAME_POOL) {
        if (std::wcscmp(name, candidate) == 0) return true;
    }
    return false;
}

} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(BunnyTests)
{
public:
TEST_METHOD(BunnyConstantsArePinnedToSpecValues) {
    Assert::IsTrue(BUNNY_COUNT_MIN == 1);
    Assert::IsTrue(BUNNY_COUNT_MAX == 2);
    Assert::IsTrue(BUNNY_HOP_SPEED_MIN == Near(22.0));
    Assert::IsTrue(BUNNY_HOP_SPEED_MAX == Near(38.0));
    Assert::IsTrue(BUNNY_BODY_RADIUS == Near(8.0));
    Assert::IsTrue(BUNNY_BODY_HEIGHT == Near(6.5));
    Assert::IsTrue(BUNNY_HEAD_RADIUS == Near(4.2));
    Assert::IsTrue(BUNNY_EAR_HEIGHT == Near(9.0));
    Assert::IsTrue(BUNNY_EAR_WIDTH == Near(2.2));
    Assert::IsTrue(BUNNY_EAR_SPACING == Near(3.0));
    Assert::IsTrue(BUNNY_LEG_LENGTH == Near(4.0));
    Assert::IsTrue(BUNNY_TAIL_RADIUS == Near(2.4));
    Assert::IsTrue(BUNNY_BODY_COLOR == 0xFF8A6A4Au);
    Assert::IsTrue(BUNNY_BELLY_COLOR == 0xFFC4A98Du);
    Assert::IsTrue(BUNNY_EAR_COLOR == 0xFF8A6A4Au);
    Assert::IsTrue(BUNNY_EAR_INNER_COLOR == 0xFFD9A0A0u);
    Assert::IsTrue(BUNNY_TAIL_COLOR == 0xFFF7F4EBu);
    Assert::IsTrue(BUNNY_EYE_COLOR == 0xFF1A1208u);
    Assert::IsTrue(BUNNY_NOSE_COLOR == 0xFF8A4040u);
    Assert::IsTrue(BUNNY_STATE_HOPPING == 0);
    Assert::IsTrue(BUNNY_STATE_GRAZING == 1);
    Assert::IsTrue(BUNNY_STATE_IDLE == 2);
    Assert::IsTrue(BUNNY_STATE_SLEEPING == 3);
    Assert::IsTrue(BUNNY_STATE_STARTLED == 4);
    Assert::IsTrue(BUNNY_HOP_DURATION == Near(0.40));
    Assert::IsTrue(BUNNY_HOP_HEIGHT == Near(8.0));
    Assert::IsTrue(BUNNY_HOP_GAP_MIN == Near(0.05));
    Assert::IsTrue(BUNNY_HOP_GAP_MAX == Near(0.20));
    Assert::IsTrue(BUNNY_GRAZE_DURATION_MIN == Near(2.5));
    Assert::IsTrue(BUNNY_GRAZE_DURATION_MAX == Near(4.5));
    Assert::IsTrue(BUNNY_IDLE_DURATION_MIN == Near(2.0));
    Assert::IsTrue(BUNNY_IDLE_DURATION_MAX == Near(4.0));
    Assert::IsTrue(BUNNY_SLEEP_DURATION_MIN == Near(6.0));
    Assert::IsTrue(BUNNY_SLEEP_DURATION_MAX == Near(12.0));
    Assert::IsTrue(BUNNY_GRAZE_PROBABILITY == Near(0.55));
    Assert::IsTrue(BUNNY_IDLE_PROBABILITY == Near(0.30));
    Assert::IsTrue(BUNNY_SLEEP_PROB == Near(0.05));
    Assert::IsTrue(BUNNY_STARTLE_RADIUS == Near(90.0));
    Assert::IsTrue(BUNNY_STARTLE_BOOST == Near(2.0));
    Assert::IsTrue(BUNNY_STARTLE_HOP_HEIGHT == Near(14.0));
    Assert::IsTrue(BUNNY_STARTLE_DURATION == Near(3.0));
    Assert::IsTrue(BUNNY_NOSE_TWITCH_FREQ == Near(6.0));
    Assert::IsTrue(BUNNY_NOSE_TWITCH_AMP == Near(0.5));
    Assert::IsTrue(BUNNY_EAR_WIGGLE_FREQ == Near(1.2));
    Assert::IsTrue(BUNNY_EAR_WIGGLE_AMP == Near(0.20));
    Assert::IsTrue(BUNNY_ZZZ_CYCLE_SEC == Near(SHEEP_ZZZ_CYCLE_SEC));
    Assert::IsTrue(BUNNY_ZZZ_RISE == Near(SHEEP_ZZZ_RISE * 0.7));
    Assert::IsTrue(BUNNY_ZZZ_SIZE_START == Near(SHEEP_ZZZ_SIZE_START * 0.7));
    Assert::IsTrue(BUNNY_ZZZ_SIZE_END == Near(SHEEP_ZZZ_SIZE_END * 0.7));
    Assert::IsTrue(sizeof(BUNNY_NAME_POOL) / sizeof(BUNNY_NAME_POOL[0]) == 12);
    Assert::IsTrue(std::wcscmp(BUNNY_NAME_POOL[0], L"Clover") == 0);
    Assert::IsTrue(std::wcscmp(BUNNY_NAME_POOL[11], L"Snowdrop") == 0);
}

TEST_METHOD(GrassGenerationProducesBunnyCountInRange) {
    for (uint64_t i = 0; i < 128; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15ull;
        Sim sim = build_grass_sim(seed);
        const int bunnies = count_kind(sim, EntityKind::Bunny);
        Assert::IsTrue(bunnies >= BUNNY_COUNT_MIN);
        Assert::IsTrue(bunnies <= BUNNY_COUNT_MAX);
    }
}

TEST_METHOD(BunniesAreGrassSceneOnly) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim_set_scene(sim, Scene::Desert);
    Assert::IsTrue(count_kind(sim, EntityKind::Bunny) == 0);
    sim_set_scene(sim, Scene::Winter);
    Assert::IsTrue(count_kind(sim, EntityKind::Bunny) == 0);
    sim_set_critter(sim, CritterKind::Bunny);
    Assert::IsTrue(count_kind(sim, EntityKind::Bunny) == 0);
}

TEST_METHOD(GeneratedBunniesHaveSpeedRange) {
    Sim sim = build_grass_sim();
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Bunny) continue;
        Assert::IsTrue(std::abs(e.vx) >= BUNNY_HOP_SPEED_MIN);
        Assert::IsTrue(std::abs(e.vx) < BUNNY_HOP_SPEED_MAX);
        Assert::IsTrue(e.rotationSpeed == Near(std::abs(e.vx)));
    }
}

TEST_METHOD(GeneratedBunniesHaveNamesInPool) {
    Sim sim = build_grass_sim();
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Bunny) continue;
        Assert::IsTrue(bunny_name_in_pool(e));
    }
}

TEST_METHOD(BunnyPRNGDrawOrderFollowsSheepAndCats) {
    Prng side;
    prng_init(side, CANONICAL_TEST_SEED ^ CRITTER_PRNG_SALT);

    Sim sim = build_grass_sim();

    const int sheepCount = prng_count(side, SHEEP_COUNT_MIN, SHEEP_COUNT_MAX);
    advance_sheep(side, sheepCount);
    const int catCount = prng_count(side, CAT_COUNT_MIN, CAT_COUNT_MAX);
    advance_cats(side, catCount);
    const int bunnyCount = prng_count(side, BUNNY_COUNT_MIN, BUNNY_COUNT_MAX);
    Assert::IsTrue(count_kind(sim, EntityKind::Bunny) == bunnyCount);

    int seen = 0;
    for (const Entity& e : sim.entities) {
        if (e.kind != EntityKind::Bunny) continue;
        const double margin = BUNNY_BODY_RADIUS + 8.0;
        const double xFrac = prng_uniform(side, 0.0, 1.0);
        const double expectedX = margin + xFrac * (Monitor1920 - 2.0 * margin);
        const uint64_t vxSign = prng_next_u64(side) & 1ull;
        const double expectedDir = vxSign != 0ull ? 1.0 : -1.0;
        const double expectedSpeed = prng_uniform(side, BUNNY_HOP_SPEED_MIN, BUNNY_HOP_SPEED_MAX);
        const uint8_t expectedName = static_cast<uint8_t>(prng_index(side,
            static_cast<uint32_t>(sizeof(BUNNY_NAME_POOL) / sizeof(BUNNY_NAME_POOL[0]))));
        Assert::IsTrue(e.x == Near(expectedX));
        Assert::IsTrue(e.vx == Near(expectedDir * expectedSpeed));
        Assert::IsTrue(e.nameIndex == expectedName);
        ++seen;
    }
    Assert::IsTrue(seen == bunnyCount);
}

TEST_METHOD(BunnyEdgeBounceFlipsDirection) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim.currentScene = Scene::Desert;
    sim.entities.clear();
    Entity e = bunny_entity(Monitor1920 - (BUNNY_BODY_RADIUS + 2.0) + 0.1, BUNNY_HOP_SPEED_MIN);
    e.stateTimer = 10.0;
    sim.entities.push_back(e);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities.front().vx < 0.0);
}

TEST_METHOD(BunnyStartleRadiusHopsAwayAndOutsideClickDoesNothing) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim.entities.clear();
    Entity e = bunny_entity(500.0, -BUNNY_HOP_SPEED_MIN);
    e.state = BUNNY_STATE_IDLE;
    e.stateTimer = 3.0;
    sim.entities.push_back(e);

    sim_apply_click(sim, click_event(500.0 - 20.0, e.y));
    Assert::IsTrue(sim.entities.front().state == BUNNY_STATE_STARTLED);
    Assert::IsTrue(sim.entities.front().vx > 0.0);
    Assert::IsTrue(sim.entities.front().stateTimer == Near(BUNNY_STARTLE_DURATION));

    Entity after = sim.entities.front();
    const double vxBefore = after.vx;
    const uint8_t stateBefore = after.state;
    sim_apply_click(sim, click_event(after.x + BUNNY_STARTLE_RADIUS + 10.0, after.y));
    Assert::IsTrue(sim.entities.front().state == stateBefore);
    Assert::IsTrue(sim.entities.front().vx == Near(vxBefore));
}

TEST_METHOD(BunnyWakesFromSleepOnStartle) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim.entities.clear();
    Entity e = bunny_entity(500.0, BUNNY_HOP_SPEED_MIN);
    e.state = BUNNY_STATE_SLEEPING;
    e.stateTimer = 10.0;
    sim.entities.push_back(e);

    sim_apply_click(sim, click_event(e.x + 10.0, e.y));

    Assert::IsTrue(sim.entities.front().state == BUNNY_STATE_STARTLED);
    Assert::IsTrue(sim.entities.front().state != BUNNY_STATE_SLEEPING);
    Assert::IsTrue(sim.entities.front().vx < 0.0);
}

TEST_METHOD(BunnyHopArcIsBounded) {
    Assert::IsTrue(bunny_hop_y_offset(0.0, false) == Near(0.0));
    Assert::IsTrue(bunny_hop_y_offset(BUNNY_HOP_DURATION, false) == Near(0.0).margin(1e-12));
    const double peak = bunny_hop_y_offset(BUNNY_HOP_DURATION * 0.5, false);
    Assert::IsTrue(peak > 0.0);
    Assert::IsTrue(peak <= BUNNY_HOP_HEIGHT);
}

TEST_METHOD(BunnyStateTransitionProbabilitiesAreStable) {
    Prng p;
    prng_init(p, CANONICAL_TEST_SEED ^ CRITTER_PRNG_SALT);
    constexpr int N = 10000;
    int graze = 0;
    int idle = 0;
    int sleep = 0;
    for (int i = 0; i < N; ++i) {
        const uint8_t state = bunny_choose_rest_state(p);
        if (state == BUNNY_STATE_GRAZING) ++graze;
        else if (state == BUNNY_STATE_IDLE) ++idle;
        else if (state == BUNNY_STATE_SLEEPING) ++sleep;
    }

    const double sleepProb = BUNNY_SLEEP_PROB;
    const double activeWeight = BUNNY_GRAZE_PROBABILITY + BUNNY_IDLE_PROBABILITY;
    const double expectedGraze = (1.0 - sleepProb) * BUNNY_GRAZE_PROBABILITY / activeWeight;
    const double expectedIdle = (1.0 - sleepProb) * BUNNY_IDLE_PROBABILITY / activeWeight;
    Assert::IsTrue(static_cast<double>(sleep) / N == Near(sleepProb).margin(0.02));
    Assert::IsTrue(static_cast<double>(graze) / N == Near(expectedGraze).margin(0.02));
    Assert::IsTrue(static_cast<double>(idle) / N == Near(expectedIdle).margin(0.02));
}

TEST_METHOD(BunnySleepProbabilityIsStable) {
    constexpr int N = 20000;
    Prng p;
    prng_init(p, CANONICAL_TEST_SEED ^ 0x1234ull);
    int sleep = 0;
    for (int i = 0; i < N; ++i) {
        if (bunny_choose_rest_state(p) == BUNNY_STATE_SLEEPING) ++sleep;
    }
    Assert::IsTrue(static_cast<double>(sleep) / N == Near(BUNNY_SLEEP_PROB).margin(0.02));
}
};
}
